astropy/io/fits/tests/test_table.py

```diff
@@ -3035,6 +3035,33 @@ def test_tolist_from_file(self):
             assert hdu.data.tolist() == [[[45, 56], [11, 3]], [[11, 12, 13], [12, 4]]]
             assert hdu.data['var'].tolist() == [[45, 56], [11, 12, 13]]
 
+    @pytest.mark.skipif('sys.maxsize < 2**32')
+    @pytest.mark.skipif('sys.platform == "win32"')
+    #@pytest.mark.hugemem
+    def test_heapsize_P_limit(self):
+        """
+        Regression test for https://github.com/astropy/astropy/issues/10812
+
+        Check if the error is raised when the heap size is bigger than what can be
+        indexed with a 32 bit signed int.
+        """
+
+        # a matrix with variable length array elements is created
+        nelem = 2**28
+        matrix = np.zeros(1, dtype=np.object_)
+        matrix[0] = np.arange(0., float(nelem+1))
+
+        col = fits.Column(name='MATRIX', format=f'PD({nelem})',
+                          unit='', array=matrix)
+
+        t = fits.BinTableHDU.from_columns([col])
+        t.name = 'MATRIX'
+
+        with pytest.raises(ValueError) as err:
+            t.writeto(self.temp('matrix.fits'))
+
+        assert "Please consider using the 'Q' format for your file." in str(err.value)
```

Comment:
you can use `pytest.raises(ValueError, match='Please consider using ...')` instead.